lib/asciidoctor/converter.rb

```diff
@@ -1,225 +1,394 @@
-        end
+    # Public: Get the Hash of Converter classes keyed by backend name. Intended for testing only.
+    def converters
+      registry.dup
+    end
+
+    private def registry
+      raise ::NotImplementedError, %(#{Factory} subclass #{self.class} must implement the ##{__method__} method)
+    end
+  end
+
+  class CustomFactory
+    include Factory
+
+    def initialize seed_registry = nil
+      if seed_registry
+        seed_registry.default = seed_registry.delete '*'
+      else
+        seed_registry = {}
       end
+      @registry = seed_registry
+    end
+
+    # Public: Unregister all Converter classes that are registered with this factory. Intended for testing only.
+    #
+    # Returns nothing.
+    def unregister_all
+      registry.clear.default = nil
+    end
 
-      def basebackend value = nil
-        if value
-          backend_info[:basebackend] = value
+    private
+
+    attr_reader :registry
+  end
+
+  # Mixed into the {Converter} module to provide the global registry of converters that are registered statically.
+  #
+  # This registery includes built-in converters for {Html5Converter HTML 5}, {DocBook5Converter DocBook 5} and
+  # {DocBook45Converter DocBook 4.5}, as well as any custom converters that have been discovered or explicitly
+  # registered. Converter registration is synchronized (where applicable) and is thus guaranteed to be thread safe.
+  module DefaultFactory
+    include Factory
+
+    private
+
+    @@registry = {}
+
+    def registry
+      @@registry
+    end
+
+    unless RUBY_ENGINE == 'opal' # the following block adds support for synchronization and lazy registration
+      public
+
+      def register converter, *backends
+        if @@mutex.owned?
+          backends.each {|backend| backend == '*' ? (@@catch_all = converter) : (@@registry = @@registry.merge backend => converter) }
         else
-          backend_info[:basebackend]
+          @@mutex.synchronize { register converter, *backends }
         end
       end
 
-      def outfilesuffix value = nil
-        if value
-          backend_info[:outfilesuffix] = value
-        else
-          backend_info[:outfilesuffix]
+      def unregister_all
+        @@mutex.synchronize do
+          @@catch_all = nil
+          @@registry = @@registry.select {|backend| PROVIDED[backend] }
         end
       end
 
-      def htmlsyntax value = nil
-        if value
-          backend_info[:htmlsyntax] = value
-        else
-          backend_info[:htmlsyntax]
+      def for backend
+        @@registry.fetch backend do
+          PROVIDED[backend] ? @@mutex.synchronize do
+            # require is thread-safe, so no reason to refetch
+            require PROVIDED[backend]
+            @@registry[backend]
+          end : catch_all
         end
       end
 
-      def supports_templates
-        backend_info[:supports_templates] = true
-      end
+      PROVIDED = {
+        'docbook45' => %(#{__dir__}/converter/docbook45),
```

Comment:
DocBook 4.5 is still a thing ? https://github.com/asciidoctor/asciidoctor/issues/3005
I believe it's the next step but maybe we can add a notice to express deprecation ?

Reply:
On the way. That will happen after this gets merged.